Input writing refactors

crisbeto opened the pull request:
Includes some refactors around how we store input data internally. The result is that:
* We duplicate less information between the `TNode` and the directive definition.
* `DirectiveDef.inputs` has a consistent type no matter how an input is configured which makes it easier to reason about.
* We need less information in order to write to a directive's input.
* We can remove the `InputTransformsFeature`.
* We can remove the `DirectiveDef.inputTransforms` field.
* We no longer need to store input flags in `TNode.inputs`.
* We store 2 fewer values in the `TNode.initialInputs` data structure.
* We don't resolve the native element for property writes in production mode. In development mode it's only used for the `ng-reflect-` attributes.

Detailed list of changes:

### refactor(core): simplify how inputs are stored in the directive definition
Currently the values in `DirectiveDef.inputs` are either strings or arrays, depending if there are flags. This makes it a bit hard to work with, because each time it's read, the consumer needs to account for both cases.

These changes rework it so the values are always an arrays.

### refactor(core): remove inputTransforms from definition 
Removes the `inputTransform` from the directive definition since this information is already available on the `inputs`.

### refactor(compiler): remove input transforms feature 
An earlier refactor made the `InputTransformsFeature` a no-op so these changes remove the code that was generating it.

### refactor(core): remove flags from TNode.inputs 
The input flags are already available on the definition so we don't need to store them on `TNode.inputs`.

### refactor(core): simplify InitialInputs data structure
Reworks the `InitialInputs` data structure to only store a public name and initial value, resulting in less memory usage and making it easier to work with.

### refactor(core): refactor(core): avoid unnecessary lookup when writing inputs
Currently we resolve the DOM node when writing inputs up-front, because it's necessary for the `ng-reflect-` attributes. Since the attributes are dev-mode-only, we can move the resolution into the function that writes them so we can avoid the resolution when it's not used.

### refactor(core): add assertion to avoid writes to directive factories
Attempting to write to directive inputs before the directive is created can lead to subtle issues that won't necessarily trigger errors. These changes add an assertion to catch such issues earlier.

crisbeto committed:
refactor(core): simplify how inputs are stored in the directive definition

Currently the values in `DirectiveDef.inputs` are either strings or arrays, depending if there are flags. This makes it a bit hard to work with, because each time it's read, the consumer needs to account for both cases.

These changes rework it so the values are always an arrays.

crisbeto committed:
refactor(core): remove inputTransforms from definition

Removes the `inputTransform` from the directive definition since this information is already available on the `inputs`.

crisbeto committed:
refactor(compiler): remove input transforms feature

An earlier refactor made the `InputTransformsFeature` a no-op so these changes remove the code that was generating it.

crisbeto committed:
refactor(core): remove flags from TNode.inputs

The input flags are already available on the definition so we don't need to store them on `TNode.inputs`.

crisbeto committed:
refactor(core): avoid unnecessary lookup when writing inputs

Currently we resolve the DOM node when writing inputs up-front, because it's necessary for the `ng-reflect-` attributes. Since the attributes are dev-mode-only, we can move the resolution into the function that writes them so we can avoid the resolution when it's not used.

crisbeto committed:
refactor(core): add assertion to avoid writes to directive factories

Attempting to write to directive inputs before the directive is created can lead to subtle issues that won't necessarily trigger errors. These changes add an assertion to catch such issues earlier.

crisbeto committed:
refactor(core): simplify InitialInputs data structure

Reworks the `InitialInputs` data structure to only store a public name and initial value, resulting in less memory usage and making it easier to work with.

crisbeto committed:
build: update bundle goldens

Updates the bundle goldens to account for the latest changes.

crisbeto commented on packages/core/src/render3/interfaces/node.ts:
````diff
@@ -793,16 +793,15 @@ export type NodeOutputBindings = Record<string, (number | string)[]>;
  * ```
  * {
- *   "publicName": [0, 'change-minified', <bit-input-flags>]
+ *   "publicName": [0, 'change-minified']
````
We're not too far off from having this be a `number[]`. The problem is that it doesn't allow us to handle re-aliased host directive inputs.

crisbeto commented:
[Passing TGP](http://test/OCL:727810502:BASE:727931679:1739825727234:6260c23a), aside from some unrelated broken targets.

pkozlowski-opensource commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -98,51 +98,40 @@ export class ComponentFactoryResolver extends AbstractComponentFactoryResolver {
-      });
+function toOutputRefArray<T>(map: DirectiveDef<T>['outputs']): ComponentFactory<T>['outputs'] {
+  const result: ComponentFactory<T>['outputs'] = [];
+  for (const publicName in map) {
```
Use `for (const key of Object.keys(obj)) { ...` or `for (const [key, value] of Object.entries(obj)) {` so we can skip `hasOwnProperty` checks?

pkozlowski-opensource commented on packages/core/test/bundling/animations-standalone/bundle.golden_symbols.json:
```diff
@@ -423,7 +423,8 @@
   "onLeave",
   "optimizeGroupPlayer",
-  "parseAndConvertBindingsForDefinition",
+  "parseAndConvertInputsForDefinition",
```
I _know_ that this is super-annoying but update to the golden symbols should be done in a commit where the actual change is made (instead of being a separate commit). This is so we can reason about commits in isolation. 

Having said all this I'm fine with squashing all the commits here into one (although your split it into individual commits is super-nice!)

pkozlowski-opensource reviewed (approved):
Very reasonable moves! 

Special kudos for having small, focused commits for each refactor - makes it so much easier to review 🎉

crisbeto committed:
refactor(core): simplify serialization of inputs and outputs

Simplifies the functions that serialize inputs/outputs for the `ComponentFactory` type.

thePunderWoman commented:
This PR was merged into the repository by commit f41437e6f5e5ad445e7b9e46e60cf3c6f41f0b74.

The changes were merged into the following branches: main, 19.1.x
